LM Studio OSS Support

rugvedS07 opened the pull request:
## Overview

Adds LM Studio OSS support. Closes #1883


### Changes
This PR enhances the behavior of `--oss` flag to support LM Studio as a provider. Additionally, it introduces a new flag`--local-provider` which can take in `lmstudio` or `ollama` as values if the user wants to explicitly choose which one to use.

If no provider is specified `codex --oss` will auto-select the provider based on whichever is running.

#### Additional enhancements 
The default can be set using `oss-provider` in config like:

```
oss_provider = "lmstudio"
```

For non-interactive users, they will need to either provide the provider as an arg or have it in their `config.toml`

### Notes
For best performance, [set the default context length](https://lmstudio.ai/docs/app/advanced/per-model) for gpt-oss to the maximum your machine can support

DocMAX commented:
please merge asap, thanks!

rugvedS07 committed:
Implement LM Studio as an OSS Provider

* Implement LM Studio as a OSS Provider
* Initial implementation of LM Studio as a OSS provider

* Cleanup lib.rs

* Save oss-provider in the config

* Code cleanup

* Auto-select running provider

* Tests for LMS Client

* Linter fixes

* Fallback to lms explicit path if lms is not bootstrapped

* Fix ordering and touch up test w/ validation

* Make sure profiles work with oss-provider

* Separate out oss and oss provider flags

* Add CTRL+C support in oss selection and change oss-provider to
local-provider in the cli arg only

* Refactor client code

* Update README and config

* Better comments in README

* Graceful handling for cancellation from oss selection screen and
warmup logic

* Make the warmup logic non-blocking

* Fix Cargo.lock

* Update config.md and linter fixes

mattjcly committed:
Wire v1/responses

mattjcly committed:
Use v1/responses JIT model loading for LMStudioClient load_model

Bauvater commented:
pls update it

rugvedS07 committed:
Fix rebasing issue

rugvedS07 committed:
Run formatter

Run formatter

rugvedS07 committed:
Formatting fix for .md

etraut-openai committed:
Merge branch 'main' into feat/lmstudio

etraut-openai committed:
Merge branch 'main' into feat/lmstudio

etraut-openai committed:
Merge branch 'main' into feat/lmstudio

etraut-openai commented on codex-rs/lmstudio/src/client.rs:
```diff
@@ -0,0 +1,397 @@
+    base_url: String,
+}
+
+const LMSTUDIO_CONNECTION_ERROR: &str = "Can't reach LM Studio. Is it running with the server on?  If you don't have it yet, try to run `lms server start` after installation https://lmstudio.ai/download";
```
Nit: Double space between sentences should be single space.

Error message is a bit cumbersome. Suggestion: "LM Studio cannot be contacted. Install from https://lmstudio.ai/download and run 'lms server start'."

etraut-openai commented on codex-rs/tui/src/lib.rs:
```diff
@@ -124,21 +129,103 @@ pub async fn run_main(
+
     let model_provider_override = if cli.oss {
-        Some(BUILT_IN_OSS_MODEL_PROVIDER_ID.to_owned())
+        if let Some(provider) = &cli.oss_provider {
```
This logic is largely duplicated between the tui and exec crates. The only difference is that the tui version falls back to an interactive menu. This duplication will create maintenance issues because override logic will need to be kept in sync between the two places. This logic should be unified in the config crate so it's implemented only once.

etraut-openai commented on codex-rs/tui/src/lib.rs:
```diff
@@ -124,21 +129,103 @@ pub async fn run_main(
     };
 
+    // When using `--oss`, let the bootstrapper pick the model based on selected provider
+    let model = if let Some(model) = &cli.model {
```
This logic is also duplicated. Look for ways to consolidate.

etraut-openai reviewed (changes requested):
Overall, this looks good to me. I left a few suggestions.

In addition to reviewing the code, I also did some manual testing and confirmed that it works as advertised with both LM Studio and Ollama config setts.

rugvedS07 committed:
Move resolve oss provider logic to config crate

rugvedS07 committed:
Move rest of oss provider code to codex-common

Making an isolated commit to move rest of the shared code to common to
avoid circular dependencies and make it easy to revert if not the
correct place
